dali/pipeline/operators/arithmetic/arithmetic_meta.h

```diff
@@ -0,0 +1,345 @@
+
+#define REGISTER_BINARY_IMPL_BACKEND(OP, EXPRESSION, BACKEND)                          \
+  template <>                                                                          \
+  template <typename L, typename R>                                                    \
+  DALI_HOST_DEV constexpr std::enable_if_t<GetOpArity(OP) == 2, binary_result_t<L, R>> \
+  arithm_meta<OP, BACKEND>::impl(L l, R r) {                                           \
+    static_assert(GetOpArity(OP) == 2,                                                 \
+                  "Registered operation arrity does not match the requirements.");     \
+    return l EXPRESSION r;                                                             \
+  }                                                                                    \
+  template <>                                                                          \
+  inline std::string arithm_meta<OP, BACKEND>::to_string() {                           \
+    return #EXPRESSION;                                                                \
+  }
+
+#define REGISTER_BINARY_IMPL(OP, EXPRESSION)               \
+  REGISTER_BINARY_IMPL_BACKEND(OP, EXPRESSION, CPUBackend) \
+  REGISTER_BINARY_IMPL_BACKEND(OP, EXPRESSION, GPUBackend)
+
+REGISTER_BINARY_IMPL(ArithmeticOp::add, +);
+REGISTER_BINARY_IMPL(ArithmeticOp::sub, -);
+REGISTER_BINARY_IMPL(ArithmeticOp::mul, *);
+REGISTER_BINARY_IMPL(ArithmeticOp::div, /);
+
+template <>
+struct arithm_meta<ArithmeticOp::mod, CPUBackend> {
+  template <typename L, typename R>
+  static constexpr std::enable_if_t<
+      std::is_integral<L>::value && std::is_integral<R>::value, binary_result_t<L, R>>
+  impl(L l, R r) {
+    return l % r;
+  }
+
+  template <typename L, typename R>
+  static constexpr std::enable_if_t<
+      !std::is_integral<L>::value || !std::is_integral<R>::value, binary_result_t<L, R>>
+  impl(L l, R r) {
+    using L_promotion = std::conditional_t<std::is_same<float16, L>::value, float, L>;
+    using R_promotion = std::conditional_t<std::is_same<float16, R>::value, float, R>;
+    return std::remainder(static_cast<L_promotion>(l), static_cast<R_promotion>(r));
+  }
+
+  static std::string to_string() {
+    return "%";
+  }
+
+  static constexpr int num_inputs = 2;
+  static constexpr int num_outputs = 1;
+};
+
+template <>
+struct arithm_meta<ArithmeticOp::mod, GPUBackend> {
+  template <typename L, typename R>
+  __device__ static constexpr std::enable_if_t<
+      std::is_integral<L>::value && std::is_integral<R>::value, binary_result_t<L, R>>
+  impl(L l, R r) {
+    return l % r;
+  }
+
+  template <typename L, typename R>
+  __device__ static constexpr std::enable_if_t<
+      (!std::is_integral<L>::value || !std::is_integral<R>::value) &&
+          (sizeof(L) < sizeof(double) && sizeof(R) < sizeof(double)),
+      binary_result_t<L, R>>
+  impl(L l, R r) {
+    return remainderf(static_cast<float>(l), static_cast<float>(r));
```

Comment:
you could merge both 'float' and 'double' cast implementation by using something like:
```
using L_promotion = std::conditional_t<sizeof(L) >= sizeof(double), double, float>;
using R_promotion = std::conditional_t<sizeof(R) >= sizeof(double), double, float>;
return remainder(static_cast<L_promotion>(l), static_cast<R_promotion>(r));
```

Reply:
remainder vs remainderf